baseapp/abci.go

```diff
@@ -192,8 +192,24 @@ func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeg
 			WithHeaderHash(req.Hash)
 	}
 
+	ctx := app.deliverState.ctx
+	if app.preBeginBlocker != nil {
+		rsp, err := app.preBeginBlocker(ctx, req)
+		if err != nil {
+			panic(fmt.Errorf("Run preBeginBlock failed, height: %d, err: %w", req.Header.Height, err))
+		}
+		// Manager skips this step if Block is non-nil since upgrade module is expected to set this params
+		// and consensus parameters should not be overwritten.
+		if rsp.ConsensusParamsChanged {
+			if cp := app.GetConsensusParams(ctx); cp != nil {
+				ctx = ctx.WithConsensusParams(cp)
```

Comment:
why we don't override when it's `nil`?

Reply:
Before migration, ctx has a non-nil ConsensusParams while app.GetConsensusParams(ctx) is nil, so there may be a problem if app.GetConsensusParams(ctx) overrides ctx ConsensusParams to nil here.

Reply:
I mean `app.GetConsensusParams()` is more authentic than `ctx.ConsensusParams()`, `ctx.ConsensusParams()` should always been an in memory cache of `app.GetConsensusParams()`.

Reply:
yes, this is after migration?
and the comment is wrong I think: "Manager skips this step if Block is non-nil since upgrade module is expected to set this params", since upgrade module can't modify it?

Reply:
yes for future use, like when people add another PreBeginBlockModule and ConsensusParamsChanged will be nil

Reply:
> rsp.ConsensusParamsChanged represents app.GetConsensusParams is non-nil

changed just means it's changed, we don't know what does it changed into, it could be a nil?

Reply:
> rsp.ConsensusParamsChanged is true from preBeginBlocker, write non-nil app ConsensusParams to context

I dont know what's wrong with it. 

> changed just means it's changed

Yes changed means true, like before migrate, when change ctx's ConsensusParams while app.GetConsensusParams still is nil.

But ConsensusParamsChanged is confusing, since we have 2 places that save cp.

Reply:
how about we just assume is the change from app.GetConsensusParams

Reply:
> how about we just assume is the change from app.GetConsensusParams

of course, `ConsensusParamsChanged` means the stuff in database is changed, that's why we need to update `ctx.ConsensusParams`.

Reply:
> write non-nil app ConsensusParams to context

there's still the question, why write *non-nil*?, rather than writing whatever returned in store?

Reply:
won't we consider how other people use it? like another PreBeginBlock which return ConsensusParamsChanged, this could be nil, and people won't know this true means new or old cp

Reply:
> won't we consider how other people use it? like another PreBeginBlock which return ConsensusParamsChanged, this could be nil, and people won't know this true means new or old cp

what could be `nil`? `app.GetConsensusParams`?

Reply:
Old means ctx.WithConsensusParams before migrate, If I am working with the OLD method of ctx consensusParams, so I return ConsensusParamsChanged true. Like write sth a XYZModule which has PreBeginBlock, it's totally valid.

Reply:
> Old means ctx.WithConsensusParams before migrate, If I am working with the OLD method of ctx consensusParams, so I return ConsensusParamsChanged true. Like write sth a XYZModule which has PreBeginBlock, it's totally valid.

it just triggers a refreshing of ctx.WithConsensusParams from the store, is that a problem?

Reply:
So do we need to force MigrationModule is the 1st to run [PreBeginBlock](https://github.com/crypto-org-chain/cosmos-sdk/blob/c25a49bf1e9ea1cf820c10db29d980e036c13c9e/types/module/module.go#L271), since now XYZModule could run before migrate

Reply:
`PreBeginBlock` should not rely on `ctx.ConsensusParams` then, if it do, do it at his own risk.

Reply:
To protect developer https://github.com/crypto-org-chain/cosmos-sdk/pull/186/commits/b6ae07f4fd4d9c6111990c7a7a4d451468243cd6